lib/Streaming/ByteString/Char8.hs

```diff
@@ -514,17 +516,13 @@ lines text0 = loop1 text0
 --  the \"lines\" had embedded newlines.
 --
 --  * @'unlines' . 'lines'@ will replace @\\r\\n@ with @\\n@.
-unlines :: Monad m => Stream (ByteStream m) m r ->  ByteStream m r
+unlines :: Monad m => Stream (ByteStream m) m r -> ByteStream m r
 unlines = loop where
   loop str =  case str of
     Return r -> Empty r
     Step bstr   -> do
       st <- bstr
-      let bs = unlines st
-      case bs of
-        Chunk "" (Empty r)   -> Empty r
-        Chunk "\n" (Empty _) -> bs
-        _                    -> cons' '\n' bs
+      cons' '\n' $ unlines st
```

Comment:
What was the issue here?

Reply:
It is explained in the commit message. Simply put, `unlines` needs to put back a newline after each substeam, unconditionally. It does not matter what "form" of empty the stream takes. They're all semantically equivalent, and the distinctions made here are fragile and dubious, why is `Chunk "" (Empty r)` somehow different from `Empty r` or from `Go $ pure (Empty r)` or `Go $ pure $ Chunk "" (Empty r)` or ...

It is all just wrong.  It is actually non-trivial to even construct examples where the original pattern matches happen, but it is possible, and they do the wrong thing.

Reply:
[ TL;DR, see the "TLDR" heading near the bottom, you can skip some of the comments in the middle on first reading if you like ]
I've found the commit that [introduced](https://github.com/haskell-streaming/streaming-bytestring/commit/a8fe89f5aeb3fd30ba85dfe5eb706d3a6caac8a2#diff-9f1312b235f4ef67d4d8608574595e4aae11d7cd75f98bdd784527c506e0d320L416-R429) the incorrect behaviour, the special cases just look like a mistake, there is no stated rationale for introducing them, and the original code was closer to being correct.

I'm essentially returning to the original implementation, but it too had a bug, it was incorrectly turning an empty stream (of bytestream lines) into a singleton stream containing one line, but that's wrong.  Empty streams must stay empty, an empty stream is not the same as a stream with a singline empty line.

Perhaps that original bug (and trying to deal with it in some way) is the reason for the complex cases in the new implementation, but both are wrong.

Basically, given a stream of lines (as `Stream (ByteStream m) m r`) in which the newline ending had been removed from each one, our job is just to put a newline after each line, which can be done by prepending a newline to the result of `unlines` on the next stream, and doing nothing when we reach the end of the outer stream, i.e. `Return r -> Empty r`.

This way:
- `Return r -> Empty r`.  No lines in => no lines out.
- The last line gets its newline ending prepended to the final `Empty r`.
- Same thing for all previous lines, they prepend a newline to the stream obtained by calling `unlines` on the remaining lines.

Because lines 525 and 526 were pattern matching the output of `unlines`, it is useful t consider the conditions under which this same `unlines` can produce either pattern in from some `remaining lines in the stream`. For that we need to peek at the [definition](https://github.com/haskell-streaming/streaming-bytestring/commit/a8fe89f5aeb3fd30ba85dfe5eb706d3a6caac8a2#diff-9f1312b235f4ef67d4d8608574595e4aae11d7cd75f98bdd784527c506e0d320L416-R429)  of `cons'`:
```haskell
cons' :: Char -> ByteStream m r -> ByteStream m r
cons' c (Chunk bs bss) | B.length bs < 16 = Chunk (B.cons (c2w c) bs) bss
cons' c cs             = Chunk (B.singleton (c2w c)) cs
```
which merges the prepended newline with the next chunk (when the stream starts immediately with a Chunk) if it is shorted than 16 bytes, and otherwise prepends a newline as a singleton.

The `unlines` prior to this PR reads:
```haskell
unlines :: Monad m => Stream (ByteStream m) m r ->  ByteStream m r
unlines = loop where
  loop str =  case str of
    Return r -> Empty r
    Step bstr   -> do
      st <- bstr
      let bs = unlines st
      case bs of
        Chunk "" (Empty r)   -> Empty r
        Chunk "\n" (Empty _) -> bs
        _                    -> cons' '\n' bs
    Effect m  -> Go (fmap unlines m)
```
- The base case is an empty stream`Return r`, it is correctly sent to the Empty bytestream `Empty r`.
- We could also an empty stream with monadic effects, such as `Effect $ return $ Return r`. This is mapped to `Go $ return $ Empty r`, which has equivalent content to the simpler base case, but notice how the special-case pattern matches would no longer be able to match this, introducing artificial distinctions between what should be equivalent streams. The placement of effects in the stream should not alter the behaviour of `unlines`.
- The next case is a stream with just a single "line" `x ... (Return r)`. Because its `st` is `Return r` (assuming no monadic effects), neither of the special cases match and it becomes `x >> Chunk "\n" (Empty r)` by falling through to the third default case.  The result can only match the second of the special case patterns, provided `x` was actually `Empty (Return r)`.  It cannot match the first special case, there is no way to produce an empty initial chunk via either `cons'` or the second pattern, or from some non-empty `x`, so the first pattern is dead code.
- The next is a two line stream `Step (y ... (Step (x ... (Return r)))`, now the extant `unlines` will potentially match the second special pattern, if `x` was `Empty (Return)` (sans monadic effects).  And in this case we get `y >> Chunk "\n" (Empty r)` without a newline appended to `y`, and perhaps again end up in the second case when `y` is just `Empty`.

#### TLDR

Bottom line, the unpatched code turns 3 explicitly `Empty` lines: `Step $ Empty $ Step $ Empty $ Step $ Empty $ Return r` into a single line instead of the expected three lines:
```
-- Setup, prepare two streams of 3 empty lines:
-- The lines of `str1` is just `Empty (Step k)`, with `k` the next line
-- The lines of `str2` are `Chunk "" (Empty (Step k))`, with `k` the next line
--
λ> import qualified Streaming.Internal as S
λ> import qualified Streaming.ByteString.Internal as Q
λ> import qualified Streaming.ByteString.Char8 as Q
λ> let eof = S.Return ()
λ> let empty1 k = S.Step $ pure k
λ> let empty2 k = S.Step $ Q.Chunk mempty $ pure k
λ> let str1 = head $ drop 3 $ iterate empty1 eof
λ> let str2 = head $ drop 3 $ iterate empty2 eof

-- Unpatched:
λ> Q.unlines str1
Chunk "\n" (Empty (()))
λ> Q.unlines str2
Chunk "" (Chunk "\n" (Chunk "\n" (Chunk "\n" (Empty (())))))

-- Patched:
λ> Q.unlines str1
Chunk "\n\n\n" (Empty (()))
λ> Q.unlines str2
Chunk "" (Chunk "\n" (Chunk "\n" (Chunk "\n" (Empty (())))))

-- Potential revision (likely not warranted):
λ> Q.unlines str1
Chunk "\n\n\n" (Empty (()))
λ> Q.unlines str2
Chunk "\n\n\n" (Empty (()))
```

Potential revision avoids the nested `Chunk "\n"` constructs when the input stream contains empty chunks as `Chunk "" $ Empty r` rather than just `Empty r`, but I am not sure this is worthwhile. The library generally tries to avoid creating empty chunks.  So  code to optimise this case (below) is probably unwarranted:
```haskell
unlines :: Monad m => Stream (ByteStream m) m r ->  ByteStream m r
unlines = loop where
  loop str =  case str of
    Step bstr   -> case bstr of
        Chunk c  cs | B.null c -> unlines $ Step cs
                    | otherwise -> bstr >>= cons' '\n' . unlines
        Go m       -> Go $ (unlines . Step) <$> m
        Empty r    -> cons' '\n' $ unlines r
    Effect m  -> Go (fmap unlines m)
    Return r -> Empty r
```

Reply:
Thanks for that detailed explanation.